units/event_log_tests.cc

```diff
@@ -0,0 +1,183 @@
+enum LogLevel { Low, Medium, High };
+
+std::string ValueToCSVString(const std::string &value) { return value; }
+
+std::string ValueToCSVString(const int64_t value) {
+  return std::to_string(value);
+}
+
+std::string ValueToCSVString(const std::vector<int64_t> &values) {
+  std::string tmp = "\"[";
+  using std::begin;
+  using std::end;
+  auto e = end(values);
+  for (auto it = begin(values); it != e; ++it) {
+    const char *delimiter = std::next(it) != values.end() ? "," : "";
+    tmp += std::to_string(*it) + delimiter;
+  }
+  tmp += "]\"";
+  return tmp;
+}
+
+struct Attribute {
+  const char *name;
+  AttributeType value_type;
+
+  template <typename T>
+  bool isa() {
+    return value_type == T::id;
+  }
+
+  // Safely casts the Attribute to the given type (T). Valid typenames should be
+  // derived from Attribute and have an AttributeType `id` field.
+  template <typename T>
+  T *cast() {
+    return isa<T>() ? static_cast<T *>(this) : nullptr;
+  };
+};
+
+// Implements the Attribute base class for each specific `AttributeType`.
+// Valid typenames: {std::string, int64_t, std::vector<int64_t>}
+// Example:
+//    AttributeImpl<int64_t, AttributeType::Int64> timestamp("timestamp", 1234);
+template <typename T, AttributeType VT>
+struct AttributeImpl : Attribute {
+  T value;
+  static constexpr AttributeType id = VT;
+
+  AttributeImpl(const char *_name, T _value)
+      : Attribute{_name, VT}, value(_value) {}
+  T &GetValue() { return value; }
+};
+
+using VectorInt64Attr =
+    AttributeImpl<std::vector<int64_t>, AttributeType::VectorInt64>;
+using StringAttr = AttributeImpl<std::string, AttributeType::String>;
+using Int64Attr = AttributeImpl<int64_t, AttributeType::Int64>;
+
+// Event represents the base struct for a loggable event. It contains the
+// event's name and the level of importance. The derived structs must define and
+// initialize their own set of attributes.
+struct Event {
+  const char *name;
+  LogLevel log_level;
+
+  Event(const char *_name, LogLevel _log_level)
+      : name(_name), log_level(_log_level) {}
+
+  // Each implementation of an Event contains attribute(s) and
+  // overrides this function to return them.
+  virtual const std::vector<Attribute *> &GetAttributes() = 0;
+  virtual size_t GetNumAttributes() = 0;
+
+  const char *GetEventName() { return name; }
+  LogLevel GetLogLevel() { return log_level; }
+};
+
+// An event for the CreateShaderModule function.
+// Example:
+//    const int64_t create_time_ns = create_end - create_start;
+//    Int64Attr timestamp("timestamp", 1234);
+//    Int64Attr shader_hash("hash", res.shader_hash);
+//    std::vector<Attribute *> attributes = {&shader_hash, &timestamp};
+//    CreateShaderModuleEvent("create_shader_module_ns", attributes,
+//    LogLevel::High);
+struct CreateShaderModuleEvent : Event {
+  std::vector<Attribute *> attributes;
+
+  CreateShaderModuleEvent(const char *_name,
+                          std::vector<Attribute *> &_attributes,
```

Comment:
I think it would be better to take make this event own all attributes. This would imply having a field per each attribute and setting them in the constructor.

Reply:
Yeah. I wanted to do this in the next steps.